src/core/thread/mle.cpp

```diff
@@ -4238,6 +4243,54 @@ uint64_t Mle::CalcParentCslMetric(const Mac::CslAccuracy &aCslAccuracy) const
 }
 #endif
 
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE
+void Mle::HandleWedAttachTimer(void)
+{
+    switch (mWedAttachState)
+    {
+    case kWedAttaching:
+        // Connection timeout
+        if (!IsRxOnWhenIdle())
+        {
+            Get<MeshForwarder>().SetRxOnWhenIdle(false);
+        }
+
+        LogInfo("Connection window closed");
+
+        mWedAttachState = kWedDetached;
+        mWakeupCallback.InvokeAndClearIfSet(kErrorFailed);
+        break;
+    default:
+        break;
+    }
+}
+
+Error Mle::Wakeup(const Mac::ExtAddress &aWedAddress,
+                  uint16_t               aIntervalUs,
+                  uint16_t               aDurationMs,
+                  WakeupCallback         aCallback,
+                  void                  *aCallbackContext)
+{
+    Error error;
+
+    VerifyOrExit((aIntervalUs > 0) && (aDurationMs > 0), error = kErrorInvalidArgs);
+    VerifyOrExit(aIntervalUs < aDurationMs * 1000, error = kErrorInvalidArgs);
+    VerifyOrExit(mWedAttachState == kWedDetached, error = kErrorInvalidState);
+
+    SuccessOrExit(error = mWakeupTxScheduler.WakeUp(aWedAddress, aIntervalUs, aDurationMs));
+
+    mWedAttachState = kWedAttaching;
+    mWakeupCallback.Set(aCallback, aCallbackContext);
+    Get<MeshForwarder>().SetRxOnWhenIdle(true);
+    mWedAttachTimer.FireAt(mWakeupTxScheduler.GetTxEndTime() + mWakeupTxScheduler.GetConnectionWindowUs());
```

Comment:
A suggestion/thought (perhaps for a future PR)?

In `Mle`, we use `mWedAttachTimer` to track the end of the requested wake-up. The `WakeupTxScheduler` also has its own timer to track wake-ups. How about we add a mechanism where `WakeupTxScheduler` tracks the end of its wake-up transmission cycles and signals back to `Mle` that the "Wakeup" is now finished? This way, we can simplify the `Mle` code and remove `mWedAttachTimer` (and possibly make `GetTxEndTime()` or `GetConnectionWindowUs()` private methods).

What do you think?

Reply:
Ok, the meaning of the `WakeupTxScheduler`'s timer would have to be overloaded to indicate the end of the connection window if `mTxTimeUs > mTxTimeEndUs`, but right, this could simplify the public interface. I think I also wanted to avoid a circular dependency between `WakeupTxScheduler` and `Mle` but this seems acceptable pattern in OT, so I can rework this as a followup. :)